pkg/cloud/throttle/throttle.go

```
@@ -172,7 +172,12 @@ func getServiceLimiterMiddleware(limiter *ServiceLimiter) middleware.FinalizeMid
 
 		out, metadata, err := handler.HandleFinalize(ctx, input)
 		smithyErr := awserrors.ParseSmithyError(err)
-		limiter.ReviewResponseV2(ctx, smithyErr.ErrorCode())
-		return out, metadata, err
+
+		if smithyErr != nil {
+			limiter.ReviewResponseV2(ctx, smithyErr.ErrorCode())
+			return out, metadata, err
+		}
+
+		return out, metadata, nil
```

@phuhung273 
Thank you for fixing this 👍 The nil check is good on `smithyErr`.

However, Can you just return the err from L173 instead of nil? If the error is not smithy error, we will not be able to unwrap it, in such cases error should be passed to next middleware which can take care of it.

---

I did it. But after looking at the function, I can see we initialize `smithyErr` right at the start https://github.com/kubernetes-sigs/cluster-api-provider-aws/blob/a2d41b76ce36323450961b8f0940c98dc321d078/pkg/cloud/awserrors/errors.go#L239-L240

So the only way for `ParseSmithyError` to return nil is when input err is nil. Then return `err` or nil are the same.

But I think your explanation is good, it can even cover the case someone changing the logic of `ParseSmithyError` in the future.

---

Yeah, it's just the pattern of AWS SDK middleware, so didn't want to change it.